Fix Setting DHO's lifetime doesn't update its entry lifetime

ekrctb opened the pull request:
- [x] Depends on ppy/osu-framework#4411
- [x] Framework update

An issue introduced in #12500:
`HitObjectLifetimeEntry`'s `LifetimeStart` and `LifetimeEnd` are hiding base properties <https://github.com/ppy/osu/blob/ea3c5aed2fccdbb6d657b4ba93b6fa5d8b6abccd/osu.Game/Rulesets/Objects/HitObjectLifetimeEntry.cs#L47>
but `PoolableDrawableWithLifetime` uses a type parameter bounded by `LifetimeEntry`.
In the `PoolableDrawableWithLifetime` implementation, `LifetimeStart` and `LifetimeEnd` refer to the base properties, not `HitObjectLifetimeEntry`'s properties.
It caused entry's lifetime to not updated when DHO's lifetime is set.

Using hiding properties as a pseudo-overriding is fragile anyway, and making it true override is the way. It requires a framework-side change.

ekrctb committed:
Add failing test showing the issue of DHO lifetime

ekrctb committed:
Use overriding instead of hiding in HitObjectLifetimeEntry

Hidden properties are used when the type is the base class. It caused issues when `DrawableHitObject` logic is factored out to `PoolableDrawableWithLifetime` because it is using the base `LifetimeEntry`, not `HitObjectLifetimeEntry`.

ekrctb committed:
Use now-public Entry.SetLifetime method

smoogipoo commented on osu.Game/Rulesets/Objects/Pooling/PoolableDrawableWithLifetime.cs:
```diff
@@ -100,11 +100,7 @@ private void setLifetime(double start, double end)
-                Entry.LifetimeStart = start;
-                Entry.LifetimeEnd = end;
-            }
+            Entry?.SetLifetime(start, end);
```
Am I right in thinking that this is a critical importance? I imagine that setting LifetimeStart or LifetimeEnd would break it, with the case:

```
KeepAlive = true
realLifetimeStart = SomeValueA
realLifetimeEnd = SomeValueB

LifetimeStart = start
realLifetimeStart = start
realLifetimeEnd = double.MaxValue (note loss of SomeValueB)
```
If I'm correct, I think this should be documented as such.

ekrctb replied:
I didn't think of that case, but I think you are right.
Not sure how to solve this issue, though.
The logic is weird as `entry.LifetimeStart = entry.LifetimeStart` is not a no-op.
I'm not even sure where to put the comment because `PoolableDrawableWithLifetime` shouldn't care about `HitObjectLifetimeEntry`-specific behavior.

smoogipoo replied:
How about splitting up `SetLifetime()` into two `SetLifetimeStart()` and `SetLifetimeEnd()`?

ekrctb committed:
Adopt framework change of LifetimeEntry

Override SetLifetimeStart/SetLifetimeEnd separately to track individual assignment. It is necessary to ensure real lifetime is not lost when lifetime is partially updated.

ekrctb committed:
Add test of HitObjectLifetimeEntry.KeepAlive behavior

ekrctb committed:
Simplify HitObjectLifetimeEntry logic a bit

smoogipoo reviewed (dismissed):
Lgtm, pending framework bump

peppy committed:
Update framework
